code/modules/holomap/holomap.dm

```diff
@@ -0,0 +1,500 @@
+
+	var/render_x_start = (region_width > 240) ? -8 : 128 - (region_width / 2)
+	var/render_y_start = (region_width > 240) ? -8 : 128 - (region_height / 2)
+
+	var/icon_to_use = 'icons/effects/holomap_parts.dmi'
+
+	for(var/x_index in region_x_start to region_x_finish)
+		for(var/y_index in region_y_start to region_y_finish)
+			var/turf/T = locate(x_index, y_index, zlevel_rendered)
+			var/area/A = get_area(T)
+			if(A.type != /area/space && A.type != /area/space/nearstation)
+				if(location_areas[A])
+					location_areas[A][2] += T
+				else
+					location_areas[A] = list(A, list(T))
+			var/W = FALSE
+			for(var/obj/structure/window/full/WW in T.contents)
+				W = TRUE
+				break
+			for(var/obj/effect/spawner/window/WW in T.contents)
+				W = TRUE
+				break
+
+			/*
+				This works by using 2x2 sprites from 'icons/effects/holomap_parts.dmi' to construct the holomap;
+				We use different colors that we will swap later on for floors and walls (BLACK-->floor;WHITE-->wall)
+				The walls can either be a full 2x2 pixel box, 1x2 pixel wall, or a 3/4 pixel corner depending on their surroundings...
+				All the following logic is to determine what icon to use.
+			*/
+			//this is what all the logic is for, setting this variable
+			var/icon_state_to_use
+			//as well as this since it modifies the wall sprite
+			var/dir_to_use = 2
+
+			//the following variables are set conditionally on the state of the turf
+			//so we can render the right sprite for this turf
+			var/neighbor_space = FALSE
+			var/list/neighbors_alien = list()
+			var/list/neighbors_solid = list()
+
+	//--------------{{{ DISCOVERY PHASE }}}------------------------------
+
+			for(var/i in GLOB.cardinal)
+				var/turf/found_turf = get_step(T, i)
+				var/area/found_area = get_area(found_turf)
+				var/found_window = FALSE
+
+				for(var/obj/structure/window/full/WW in found_turf.contents)
+					found_window = TRUE
+					break
+				for(var/obj/effect/spawner/window/WW in found_turf.contents)
+					found_window = TRUE
+					break
+
+				for(var/j in GLOB.diagonals)
+					var/turf/found_s_turf = get_step(T, j)
+					var/area/found_s_area = get_area(found_s_turf)
+					if(found_s_area == /area/space || found_s_area == /area/space/nearstation || isspaceturf(found_s_turf))
+						neighbor_space = TRUE
+						break
+
+				if(neighbor_space || found_area == /area/space || found_area == /area/space/nearstation || isspaceturf(found_turf))
+					neighbor_space = TRUE
+					break
+
+				if(found_area != A || locate(/obj/machinery/door/airlock) in found_turf)
+					neighbors_alien += i
+					continue
+
+				if(iswallturf(found_turf) || istype(found_turf, /turf/simulated/mineral) || found_window)
+					neighbors_solid += i
+
+	//-----------------{{{ RENDERING PHASE }}}----------------------------
+
+			if(neighbor_space || length(neighbors_solid) >= 3)
+				if(iswallturf(T) || istype(T, /turf/simulated/mineral) || W)
```

Comment:
cache this check in a variable `iswallturf(T) || istype(T, /turf/simulated/mineral) || W` so we can reuse it in the next check

Reply:
you mean like assigning it as a boolean, and setting again and again for each iteration?

Reply:
Still don't get your intent with this one, mind explaining it to me?

Reply:
so, since our internal loop is going to get call 65 thousand times, it means that every small proc call or operation we can cut out or optimize is potentially a significant chunk of a tick we're gonna save over the entire execution of the proc. If we do enough we can probably even shave seconds off of the init time for your holomaps.

looking back at it, I realize that we're only ever gonna perform this check once per iteration, however we can still optimize it by move the `W` check to the front, so if that's true we can skip both of the istypes.

I recommend spending an hour just going through this proc and finding places where you can minimize the amount of operations per iteration. At some point I'm gonna go through and check your bitwise operations aswell and potentially look into finding a better method of checking stuff.